File: 2906-construct-product-matrix/2906-construct-product-matrix.java
class Solution {
    public int[][] constructProductMatrix(int[][] grid) {
        int n = grid.length, m = grid[0].length;

        var ans = new int[n][m];
        for (int i = 0; i < n; ++i)
            Arrays.fill(ans[i], 1);

        long left = 1, right = 1;
        final int MOD = 12345;

        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                ans[i][j] = (int) (ans[i][j] * left) % MOD;
                left = (left * grid[i][j]) % MOD;
            }
        }

        for (int i = n - 1; i >= 0; --i) {
            for (int j = m - 1; j >= 0; --j) {
                ans[i][j] = (int) (ans[i][j] * right) % MOD;
                right = (right * grid[i][j]) % MOD;
            }
        }

        return ans;
    }
}

File: 2906-construct-product-matrix/2906-construct-product-matrix.py
class Solution:
    def constructProductMatrix(self, grid: List[List[int]]) -> List[List[int]]:
        n, m = len(grid), len(grid[0])
        ans = [[1 for _ in range(m)] for _ in range(n)]

        left, right = 1, 1
        MOD = 12345

        for i in range(n):
            for j in range(m):
                ans[i][j] = (ans[i][j] * left) % MOD
                left = (left * grid[i][j]) % MOD
        

        for i in range(n - 1, -1, -1):
            for j in range(m - 1, -1, -1):
                ans[i][j] = (ans[i][j] * right) % MOD
                right = (right * grid[i][j]) % MOD

        return ans